src/octoprint/plugins/gcodeviewer/static/js/viewer/reader.js

```diff
@@ -117,41 +125,39 @@ GCODE.gCodeReader = (function () {
         return {layer: layer, cmd: cmd};
     };
 
-    var sanitizeModel = function (m) {
+    var cleanModel = function (m) {
         if (!m) return [];
-        return _.filter(m, function (layer) {
-            return !!layer;
-        });
-    };
 
-    var purgeEmptyLayers = function (m) {
-        return _.filter(m, function (layer) {
-            return (
-                !!layer &&
-                layer.length > 0 &&
-                _.find(layer, function (cmd) {
-                    return cmd && cmd.extrude;
-                }) !== undefined
-            );
-        });
+        var result = [];
+        rendererEmptyLayers = [];
+        rendererPercentageByLayer = [];
+        for (var i = 0; i < m.length; i++) {
+            if (!m[i]) continue;
+            if (gCodeOptions["purgeEmptyLayers"] && emptyLayers[i]) continue;
+
+            result.push(m[i]);
+            rendererPercentageByLayer.push(percentageByLayer[i]);
+            if (emptyLayers[i]) rendererEmptyLayers[result.length - 1] = true;
+        }
+        return result;
     };
 
     var rebuildLayerPercentageLookup = function (m) {
         var result = [];
-
-        if (m && m.length > 0) {
-            for (var i = 0; i < m.length - 1; i++) {
-                // start is first command of current layer
-                var start = m[i].length ? m[i][0].percentage : -1;
-
-                var end = -1;
-                for (var j = i + 1; j < m.length; j++) {
-                    // end is percentage of first command that follows our start, might
-                    // be later layers if the next layer is empty!
-                    if (m[j].length) {
-                        end = m[j][0].percentage;
-                        break;
-                    }
+        for (var i = 0; i < m.length - 1; i++) {
```

Comment:
If I remember correctly, during testing of #4430 there were sometimes cases where `m` would still be unset/empty due to a race condition during loading. That is no longer being captured here. Can you please verify that it is indeed no longer needed and add back if unsure?

Reply:
Sure, I'll add it back. 

I have never seen any errors on this, though.